inet/vibe/inet/url.d

```diff
@@ -19,11 +19,75 @@ import std.exception;
 import std.string;
 import std.traits : isInstanceOf;
 import std.ascii : isAlpha, isASCII, toLower;
-import std.uri: encode;
+import std.uri: decode, encode;
 
 import core.checkedint : addu;
 
 
+/** Parses a user-entered URL with relaxed parsing rules.
+
+	Unlike `URL.parse`, this allows the URL to use special charaters as part of
+	the host name and path, automatically emplying punycode or percent-encoding
+	to convert this to a valid URL.
+
+	Also, relative URL references are supported by supplying a base URL.
+
+	Params:
+		url = String representation of the URL
+		base_url = Base
+*/
+URL parseUserURL(string url, URL base_url = URL.init)
+{
+	// test if this is a URL or a relative (path) reference
+	auto sidx = url.indexOf(':');
+	if (sidx < 0) {
+		// TODO: handle query strings and anchors
+
+		if (url.representation.canFind('%')) {
+			try url.decode();
+			catch (Exception e) url = url.encode();
+		} else url = url.encode();
+
+		auto path = InetPath(url);
+
+		if (base_url != URL.init) {
+			if (path.absolute)
+				base_url.path = path;
+			else if (base_url.path.endsWithSlash)
+				base_url ~= path;
+			else base_url.path = base_url.path.parentPath ~ path;
+			return base_url;
+		} else {
+			if (!path.absolute) path = InetPath.fromTrustedString("/") ~ path;
+			return URL("file", null, 0, path);
+		}
+	}
+
+	return URL(url, false);
+}
+
+unittest {
```

Comment:
I'm not very found of moving those unittests. They test different functions / part of the code.
The previous unittests could be moved to documented unittests for the ctor.

Reply:
Do they test different parts, though?

Reply:
Yes and no. Yes because the encoding logic is mixed with the rest of the ctor parsing because, AFAIR, it has to be.
No because a machine-provided URL should be its `URL` representation, so in that sense the support in the constructor is "part of" `parseUserURL`. Still, unittests for the encoding itself should be next to the code, while documented unittests should be next to the API.

Reply:
Well, as soon as we go further with this, we'll have to separate the parsing logic completely. Right now it kind of makes sense to share the logic, but eventually there will be too many special rules in the `parseUserURL` path. IMO we could just remove the `encoded == false` path from the constructor and instead duplicate it to `parseUserURL` directly.